controller/uninstall_controller.go

```diff
@@ -0,0 +1,468 @@
+	daemonSetInformer v1beta2.DaemonSetInformer,
+) *UninstallController {
+	c := &UninstallController{
+		namespace:                 namespace,
+		force:                     force,
+		ds:                        ds,
+		stopCh:                    stopCh,
+		volumeInformerSynced:      volumeInformer.Informer().HasSynced,
+		engineInformerSynced:      engineInformer.Informer().HasSynced,
+		replicaInformerSynced:     replicaInformer.Informer().HasSynced,
+		engineImageInformerSynced: engineImageInformer.Informer().HasSynced,
+		nodeInformerSynced:        nodeInformer.Informer().HasSynced,
+		daemonSetInformerSynced:   daemonSetInformer.Informer().HasSynced,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 1*time.Second),
+			&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+		), "longhorn-uninstall"),
+	}
+
+	volumeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineInformer.Informer().AddEventHandler(c.controlleeHandler())
+	replicaInformer.Informer().AddEventHandler(c.controlleeHandler())
+	engineImageInformer.Informer().AddEventHandler(c.controlleeHandler())
+	nodeInformer.Informer().AddEventHandler(c.controlleeHandler())
+	daemonSetInformer.Informer().AddEventHandler(c.controlleeHandler())
+	return c
+}
+
+func (c *UninstallController) controlleeHandler() cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueControlleeChange() },
+		UpdateFunc: func(old, cur interface{}) { c.enqueueControlleeChange() },
+		DeleteFunc: func(obj interface{}) { c.enqueueControlleeChange() },
+	}
+}
+
+func (c *UninstallController) enqueueControlleeChange() {
+	c.queue.AddRateLimited("uninstall")
+}
+
+func (c *UninstallController) Run() error {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	if !controller.WaitForCacheSync("longhorn uninstall", c.stopCh,
+		c.volumeInformerSynced, c.engineImageInformerSynced,
+		c.nodeInformerSynced, c.daemonSetInformerSynced) {
+
+		return errors.New("Failed to sync informers")
+	}
+
+	if err := c.checkPreconditions(); err != nil {
+		c.stop()
+		return err
+	}
+
+	startTime := time.Now()
+	logrus.Infof("Uninstalling...")
+	defer func() {
+		if c.stopErr == nil {
+			fields := logrus.Fields{"runtime": time.Now().Sub(startTime)}
+			logrus.WithFields(fields).Infof("Uninstallation completed.")
+		}
+	}()
+	go wait.Until(c.worker, time.Second, c.stopCh)
+
+	<-c.stopCh
+	return c.stopErr
+}
+
+func (c *UninstallController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *UninstallController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.uninstall()
+	c.handleErr(err, key)
+
+	return true
+}
+
+func (c *UninstallController) handleErr(err error, key interface{}) {
+	if err == nil {
+		c.queue.Forget(key)
```

Comment:
Log a warning here. Otherwise, the key will be dropped silently.

Reply:
Since we discard all errors, this will be way too chatty. I commented above about the issue distinguishing worker errors from worker done (until next change event).